[PhpUnitBridge] Add `enum_exists` mock

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | _N/A_
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/17538

alexandre-daubois committed:
[PhpUnitBridge] Add `enum_exists` mock

stof commented on src/Symfony/Bridge/PhpUnit/Tests/ClassExistsMockTest.php:
```diff
@@ -117,3 +139,11 @@ trait ExistingTrait
 {
 }
+
+enum ExistingEnum
```
those must be placed in a separate fixture file that gets required only on PHP 8.1+. Otherwise, the test file will trigger a syntax error on PHP 8.0 and older.

alexandre-daubois replied:
Thank you, updated!

nicolas-grekas commented on src/Symfony/Bridge/PhpUnit/ClassExistsMock.php:
```diff
@@ -49,6 +49,17 @@ public static function trait_exists($name, $autoload = true)
+        $name = ltrim($name, '\\');
+
+        if (\PHP_VERSION_ID < 80100) {
+            throw new \LogicException('"enum_exists" cannot be used with a version of PHP earlier than 8.1.');
```
yes it can, there is a polyfill for that!

derrabus commented:
Question: Can I use this mocking mechanism to test code that looks like this?

```php
if (class_exists(SomeClass::class) && !enum_exists(SomeClass::class)) {
    // do something
}
```

This is a common pattern if you want to check to make sure a class is actually a class because an enum would also pass a `class_exists()` check. See https://3v4l.org/5qfuI

alexandre-daubois commented:
I actually don't have the use case. This contribution idea came to me when crawling the bridge documentation 😄 I updated the code so there's a diff between classes and enums. I added `withMockedEnums` that fills a new `$enums` properties, as well as `$classes`. This way, `class_exists` will return true on enums 👍

ro0NL commented on src/Symfony/Bridge/PhpUnit/ClassExistsMock.php:
```diff
@@ -18,16 +18,29 @@ class ClassExistsMock
+    public static function withMockedEnums(array $enums)
+    {
+        self::$enums = $enums;
+        self::$classes += $enums;
```
im not sure i understand withMockedEnums, as symbol names are unique (at the time we decided to call symbols classes, as it's a more recognizable term)

ro0NL replied:
i understand we return true for all *_exists functions, when given a specific type of symbol

but that's still the case now with writing to $classes AND $enums :thinking: 

if we make this distinction i'd prefer to do it all for all symbol types

edit: AH i figured php returns true for enums with class_exists() (but still not with interface/trait_exists) https://github.com/symfony/symfony/pull/48516#issuecomment-1353261478

derrabus commented on src/Symfony/Bridge/PhpUnit/Tests/EnumExistsMockTest.php:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+
+enum ExistingEnum
+{
+}
+
+enum ExistingEnumReal
+{
+}
```
This will cause a parse error on PHP < 8.1. If we want to be able to run the tests on lower PHP versions, those two enums have to be moved into separate files.

alexandre-daubois replied:
Fixed, thanks!

derrabus commented on src/Symfony/Bridge/PhpUnit/composer.json:
```diff
@@ -22,7 +22,8 @@
         "symfony/deprecation-contracts": "^2.5|^3.0",
-        "symfony/error-handler": "^5.4|^6.0"
+        "symfony/error-handler": "^5.4|^6.0",
+        "symfony/polyfill-php81": "^1.27"
```
Why do we need this dependency?

alexandre-daubois replied:
Because of this: https://github.com/symfony/symfony/pull/48516#discussion_r1049744618 😄

derrabus replied:
Well, okay, it's just a dev dependency, so I guess it's fine.

nicolas-grekas commented:
Thank you @alexandre-daubois.
